### PR TITLE
Improvements to XMLProperties

### PR DESCRIPTION
**Pull request by guusdk**
The changes in this PR are motivated by observed thread contention in a high-volume environment. That's where the fixes in the first two commits stem from. I've added some minor fixes in subsequent commits.

I started off by using a ConcurrentMap to fix OF-2377, but that didn't quite cover all cases (in particular with protecting access to the 'document' field of this class). Instead, I've opted to go with a ReadWriteLock. This probably adds a bit of overhead, but should be more thread-safe.

One concession that I had to make for being able to use a ReadWriteLock, is to move all modifications that happened while a 'read' lock is acquired, outside of the lock (as ReentrantReadWriteLock guarantees that trying to obtain a 'write' lock while a 'read' lock is being held will fail). I've marked one operation that applies a change to an XML property without holding a 'write' lock. Triggering that code should be rare enough for me to believe that this won't be a practical problem.

### CODE REVIEW

**Comment by GregDThomas** [comment]
Will look at this later - but am slightly curious as to the how it came about. My fading memory was that the XML file was pretty much only used to bootstrap the app until it could read the DB, so I am surprised that there was any contention at all. 

Note that I'm saying it's unnecessary, I'm sure it is, but don't know exactly why!

**Comment by guusdk** [comment]
Usage of A SystemProperty that was not backed by an actual database entry which will attempt a lookup of an XMLProperty before eventually returning the default value.

**Comment by GregDThomas** [comment]
> Usage of a SystemProperty that was not backed by an actual database entry which will attempt a lookup (every time) of an XMLProperty before eventually returning the default value.

That seems like an odd choice I made! 🤷

**Comment by guusdk** [comment]
This probably isn't you. It might not be in SystemProperties directly, but a side-effect of a call that it does to an older API.

File: xmppserver/src/main/java/org/jivesoftware/util/XMLProperties.java [hunk 1]
```diff
@@ -182,43 +168,57 @@ public synchronized String getProperty(String name) {
      * @return the value of the specified property.
      */
     public synchronized String getProperty(String name, boolean ignoreEmpty) {
```
Comment (GregDThomas) [inline xmppserver/src/main/java/org/jivesoftware/util/XMLProperties.java:170]: Still `synchronized` ?

**Comment by GregDThomas** [comment]
The code seems fine - no quibbles with the concession over the locks. I did find one `synchronized` method that needs the qualifier removing - a quick search may turn up others I've missed too.

The answer my next question is that this is probably an edge case that us used during an install or similar, but I am wondering if a better approach would be simply read in the whole XML file in to a Map of properties, and return values direct from the Map - a ConcurrentHashMap would mean no locking would be required. The Map & File can then be updated in any write statements.

**Review by GregDThomas — APPROVED** [review]

File: xmppserver/src/main/java/org/jivesoftware/util/XMLProperties.java (hunk 1 above)
Comment (guusdk) [reply xmppserver/src/main/java/org/jivesoftware/util/XMLProperties.java:170]: hargh! I CTRL-F'ed for that! Good catch, thanks!

**Comment by guusdk** [comment]
> The code seems fine - no quibbles with the concession over the locks. I did find one `synchronized` method that needs the qualifier removing - a quick search may turn up others I've missed too.
> 
> The answer my next question is that this is probably an edge case that us used during an install or similar, but I am wondering if a better approach would be simply read in the whole XML file in to a Map of properties, and return values direct from the Map - a ConcurrentHashMap would mean no locking would be required. The Map & File can then be updated in any write statements.

I think that this is good improvement, but I'd like to push that to a major release. I'm dreading some weird edge-case that's broken by this appoach.

**Comment by Fishbowler** [comment]
This is fundamental stuff, and the tests pass which gives confidence. Have you got any real world testers that can help validate this? Given the database activity, would this be a candidate for the new DB tests you'll introduce with #1990, or is that overkill? Moreover: how can I help get this finished?

**Comment by guusdk** [comment]
>

**Comment by guusdk** [comment]
> Have you got any real world testers that can help validate this?

This now has been ported to a custom fork of Openfire, where it helped reducing the startup time for Openfire from ~1h15 minutes to ~30 minutes (this wasn't the only change to go in, and their use of Properties is _very_ peculiar - but still).

> Given the database activity, would this be a candidate for the new DB tests you'll introduce with #1990, or is that overkill?

I'm not going to claim that unit tests are bad to have (feel free to add some more :smile:), but the main point of this PR is to improve performance when there is _no_ database content. Unit tests aren't really suitable for performance-measuring things. It might be good to verify behavior more, but I do think that quickly gets a bit excessive for this PR.

> how can I help get this finished?

Hit the green button?
